tfexec/version.go

```diff
@@ -0,0 +1,78 @@
+package tfexec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+)
+
+// Version returns structured output from the `terraform version` command including both the Terraform CLI version
+// and any initialized provider versions.
+func (tf *Terraform) Version(ctx context.Context) (tfVersion *version.Version, providerVersions map[string]*version.Version, err error) {
+	// TODO: support re-invocation of this?
```

Comment:
I think `Version()` should always invoke `terraform version` - `tfexec` is intended as a way to invoke `terraform` CLI commands as transparently as possible.

If the consumer wants the cached versions, they can read the fields from the `Terraform` struct.

Reply:
I agree with Katy here, I think that any caching responsibility should be up to the consumer, even if it's consumer within `terraform-exec` itself. e.g. I assume at some point we may need to cache and/or transparently expose the version here https://github.com/hashicorp/terraform-exec/blob/22dfd359c0f8d72deaecb08a0294851c4f622174/tfinstall/tfinstall.go#L133-L136

but any such caching/exposure decision should be made within that package, keeping `tfexec` itself simple and transparent.

Reply:
Yeah I agree (about re-invocation at least, I dislike a field for this, will put a new commit together with thoughts).

Especially now that this is more info than just the tf version (provider versions), re-invocation is much more useful.